docs/ConstPropagationPass.md

```diff
@@ -0,0 +1,266 @@
+}
+```
+
+## Remark 
+
+ONNXConstantOp uses MLIR DenseElementsAttr to store constant values. It is
+important to note that, once a DenseElementsAttr is created, it is alive and 
+consumes memory until the end of compilation. In [Example](#example), all the
+three DenseElementsAttrs in the three ONNXConstantOps exist until the end of
+compilation. Especially, two intermediate DenseElementsAttrs in the two
+ONNXConstantOps produced by folding the two ONNXAddOps also exist. For a
+practice model, the number of intermediate DenseElementsAttrs will increase
+quickly, which lead to a large memory footprint during compilation. 
+
+To avoid creating too many DenseElementsAttrs for intermediate ONNXConstant ops
+during `--constprop-onnx`, we design a mechanism that dynamically allocates and
+deallocates buffers for intermediate ONNXConstantOps and only creates
+DenseElementsAttr for the final results of constant propagation.
+
+In particular, we maintain a buffer pool internally. When an ONNXConstantOp
+is reached at the first time, we read its DenseElementsAttr and store
+data to an array buffer in the pool. A unique buffer ID is used to map the
+ONNXConstantOp to the buffer. All constant computations are then done on array
+buffers, which is to avoid creating DenseElementsAttr for intermediate
+ONNXConstantOps. Buffers are automatically freed if they are not used.
+
+We provide three helper functions to use when working with buffers:
+1. `getArrayFromAttributeOrBuffer(PatternRewriter &rewriter, Operation *op)`
+   - create a buffer from a dense attribute at the first time we reach the
+     const 'op' and add the buffer to the buffer pool, or
+   - get the buffer from the buffer pool if it was created.
+2. `createConstantOpAndStoreBufferPtr(..., char *buffer)`
+   - create a new ONNXConstantOp using the given buffer, and
+   - add the buffer to the buffer pool.
+3. `allocateBufferFor(Value value, bool useMaxSize = false)`
+   - create a new buffer whose size is obtained from the type of 'value'.
+
+Note that:
+  - The buffers in the buffer pool will be automatically freed. Users don't
+    need to take care about that.
+  - If we create a buffer and do not put it on the buffer pool, please
+    make sure that it is correctly freed. We don't recommend to do this.
+    
+## Write rules for constant propagation
+
+We use MLIR declarative rewriting rules (DRR) to write patterns for constant
+propagation. The DRR definition used for defining patterns is shown below:
+```
+class Pattern<
+   dag sourcePattern,
+   list<dag> resultPatterns,
+   list<dag> additionalConstraints = [],
+   dag benefitsAdded = (addBenefit 0)
+>;
+```
+
+More information about DRR can be found [here](https://mlir.llvm.org/docs/DeclarativeRewrites/).
+
+There is a limitation in writing DRRs for `--constprop-onnx` pass so that the
+memory footprint is minimized, that is:
+- Never use ONNXConstant in the result patterns of a DRR, because this
```

Comment:
you mean ONNXConsantOp?

Could we not to a custom builder that takes a buffer as input and then use one in the rules?

Reply:
Could we not to a custom builder that takes a buffer as input and then use one in the rules?

Good idea. I need to think more about how to associate it with the buffer pool. The buffer pool is not global now. It's a big change to make it global across multiple passes, but a huge benefit overall.

Reply:
np, it works as is, just trying to see if we could fit it in the scheme of ops.